### DISCUSSION TITLE
Runtime Warning for MJX

### DISCUSSION THREAD

**Discussion by Lhy6900** [google-deepmind/mujoco#2226] [Asking for Help]
### Intro

Hi!

I am a student and I am using MuJoCo for my research on Robotics.

### My setup

MJX 3.2.4
MUJOCO 3.2.4
Ubuntu 22.04

### My question

When I defined the step function in the RL environment, a warning appeared during JIT compilation and execution, so I used `jax.debug.print` in the code to find the source of the problem. I found that the problem occurred in the step function of mjx.The code and the warning are attached below.

Here is my code.
```
def step():
    ....
    jax.config.update("jax_disable_jit", False)
    pipeline_state = state.pipeline_state
    pipeline_state = pipeline_state.replace(xfrc_applied=xfrc_applied)
    state = state.replace(pipeline_state = pipeline_state)
    jax.debug.print('flag1')
    jax.config.update("jax_disable_jit", True)
    jax.config.update("jax_disable_jit", False)

    target_torque = jp.clip(target_torque, -1000, 1000)
    pipeline_state = self.pipeline_step(state.pipeline_state, target_torque)
    jax.debug.print('flag2')
    jax.config.update("jax_disable_jit", False)
    x, xd = pipeline_state.x, pipeline_state.xd

    done |= positive_transpos(x.pos[0])
    jax.debug.print('flag3 done')
    jax.config.update("jax_disable_jit", True)
    .....
```
Here is the warning.
```
flag1
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
/home/.conda/envs/jax/lib/python3.9/site-packages/jax/_src/interpreters/xla.py:155: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
flag2
flag3 done
```
This warning will only appear in the first two steps and will not appear again afterwards. Considering that the `mjx step` has no connection with the code, does this mean that the problem comes from the xml? How can I determine the source of the problem later?

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by Stocko-2073** [comment]
Did you find a solution to this? I'm facing the same issue..

**Comment by pran-d** [comment]
Hi @yuvaltassa 
I'm facing the same issue while trying to run the following code:

```
mj_model = mujoco.MjModel.from_xml_path(xml) 
mj_data = mujoco.MjData(mj_model)
mjx_model = mjx.put_model(mj_model)
mjx_data = mjx.put_data(mj_model, mj_data)
self.rng = jax.random.PRNGKey(0)
self.rng = jax.random.split(self.rng, 1000)
batch = jax.vmap(lambda rng: mjx_data.replace(qpos=jax.random.uniform(rng, (19,))))(self.rng)
jit_step = jax.jit(jax.vmap(mjx.step, in_axes=(None, 0)))
batch = jit_step(mjx_model, batch)
print(batch.qpos)
```
The code is based on MJX tutorials, and the .xml itself works normally when I use it on MuJoCo without MJX. Is there anything we should avoid in our .xml that is allowed in MuJoCo but can raise such errors in MJX?

This is the error that I encounter:
```
/anaconda3/envs/mujoco/lib/python3.12/site-packages/jax/_src/interpreters/xla.py:132: RuntimeWarning: overflow encountered in cast
  return np.asarray(x, dtypes.canonicalize_dtype(x.dtype))
```

**Reply by kevinzakka** [reply]
This is a benign warning, you can ignore it.

**Reply by pran-d** [reply]
@kevinzakka 
Is the call to `mjx.step` known to take so long to run? I have waited for over 5 minutes but the function still hasn't returned.

**Reply by ayushdhingra02** [reply]
> This is a benign warning, you can ignore it.

@kevinzakka whenever this warning comes the code seems to stop proceeding further.. its shows the cell is running but it is running forever in colab notebook

this is the part of code from tutorial notebook:
 
# initialize the state
state = jit_reset(jax.random.PRNGKey(0))
rollout = [state.pipeline_state]

# grab a trajectory
for i in range(10):
  ctrl = -0.1 * jp.ones(env.sys.nu)
  state = jit_step(state, ctrl)
  rollout.append(state.pipeline_state)

media.show_video(env.render(rollout, camera='side'), fps=1.0 / env.dt)

i am running the same tutorial code for my custom humanoid xml model and is giving this same warning . this cell keeps on running forever without showing output and no other errors

**Reply by kevinzakka** [reply]
@pran-d If it's taking too long, then it means you need to tweak the MJX solver parameters, or collision meshes. What's your setup exactly?

**Reply by kevinzakka** [reply]
@ayushdhingra02 What GPU are you using? Is this in Colab?

**Reply by ayushdhingra02** [reply]
@kevinzakka yep.. that's on colab

**Comment by flimdejong** [comment]
I have this issue as well. It runs, but takes about 50 seconds for a very simple environment. Is this normal?

**Reply by kevinzakka** [reply]
@flimdejong I think 50 seconds sounds normal (again it depends on the environment). That being said, I think there's been a regression in JAX and I've noticed that all the playground envs are now jitting a bit slower.
